fem/bilininteg_diffusion.cpp

```diff
@@ -212,6 +208,435 @@ void DiffusionIntegrator::AssemblePA(const FiniteElementSpace &fes)
                     coeff, pa_data);
 }
 
+
+template<int T_D1D = 0, int T_Q1D = 0>
+static void PADiffusionDiagonal2D(const int NE,
+                                  const Array<double> &b,
+                                  const Array<double> &g,
+                                  const Vector &d,
+                                  Vector &y,
+                                  const int d1d = 0,
+                                  const int q1d = 0)
+{
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   MFEM_VERIFY(D1D <= MAX_D1D, "");
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   auto B = Reshape(b.Read(), Q1D, D1D);
+   auto G = Reshape(g.Read(), Q1D, D1D);
+   // note different shape for D, this is a (symmetric) matrix,
+   // we only store necessary entries
+   auto D = Reshape(d.Read(), Q1D*Q1D, 3, NE);
+   auto Y = Reshape(y.ReadWrite(), D1D, D1D, NE);
+   MFEM_FORALL(e, NE,
+   {
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+      constexpr int MD1 = T_D1D ? T_D1D : MAX_D1D;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      // gradphi \cdot Q \gradphi has four terms
+      double QD0[MQ1][MD1];
+      double QD1[MQ1][MD1];
+      double QD2[MQ1][MD1];
+      double QD3[MQ1][MD1];
+      for (int qx = 0; qx < Q1D; ++qx)
+      {
+         for (int dy = 0; dy < D1D; ++dy)
+         {
+            QD0[qx][dy] = 0.0;
+            QD1[qx][dy] = 0.0;
+            QD2[qx][dy] = 0.0;
+            QD3[qx][dy] = 0.0;
+            for (int qy = 0; qy < Q1D; ++qy)
+            {
+               const int q = qx + qy * Q1D;
+               const double D0 = D(q,0,e);
+               const double D1 = D(q,1,e);
+               const double D2 = D(q,2,e);
+               QD0[qx][dy] += B(qy, dy) * B(qy, dy) * D0;
+               QD1[qx][dy] += B(qy, dy) * G(qy, dy) * D1;
+               QD2[qx][dy] += G(qy, dy) * B(qy, dy) * D2;
```

Comment:
On line 258,  ``QD2[qx][dy] += G(qy, dy) * B(qy, dy) * D2;`` should be ``QD2[qx][dy] += G(qy, dy) * B(qy, dy) * D1;``

Reply:
Nice catch! I've corrected this. (This suggests we need a test with non-constant coefficients...)

Reply:
So the PA code so far only supports constant coefficients. I've tried making the fichera test use a curved mesh (ie `fichera-q3.mesh`) which should be sort of equivalent here, but it doesn't make any difference. Any suggestions?

Reply:
The fichera meshes are 3D, you want a 2D mesh to test this, e.g. `star-q2.mesh`.

Reply:
Once branch https://github.com/mfem/mfem/pull/982 is in, we should have support for non-constant coefficients. A test could be added once the PR is merged.

Reply:
@artv3, you probably meant #1107, #982 is this PR.

Reply:
Oops sorry, yes https://github.com/mfem/mfem/pull/1107